src/transformers/generation_utils.py

```diff
@@ -3446,3 +3766,152 @@ def top_k_top_p_filtering(
         )
 
     return logits
+
+
+# ========== utils for contrastive search decoding method ========= #
```

Comment:
Those utils would probably be best in a submodule (like we have `generation_beam_search` or `generation_beam_constraint`) to avoid `generation_utils` being too big.